python/tvm/relay/qnn/op/legalizations.py

```diff
@@ -433,15 +443,15 @@ def _qnn_conv2d_legalize_arm_cpu(attrs, inputs, types):
         attrs["groups"],
     )
     use_int8_on_arm = (not is_depthwise) and is_aarch64_arm() and attrs["data_layout"] == "NHWC"
-    if use_int8_on_arm or is_fast_int8_on_arm():
+    if use_int8_on_arm or is_fast_int8_on_arm() or is_cortexm_arm():
         return helper_change_dtypes_to_be_same(attrs, inputs, types, relay.qnn.op.conv2d)
     return helper_no_fast_int8_hw_legalization(attrs, inputs, types, relay.nn.conv2d)
```

Comment:
I think this is actually the same as an issue awhile back (https://github.com/apache/tvm/issues/8717#issuecomment-976400930), instead of special casing with `is_cortexm` you should be able to instead use something like:

```suggestion
    has_asimd = (is_aarch64_arm() or "+neon" in target.mattr)
    if has_asimd and not is_fast_int8_on_arm():
        return helper_no_fast_int8_hw_legalization(attrs, inputs, types, relay.nn.conv2d)
    return helper_change_dtypes_to_be_same(attrs, inputs, types, relay.qnn.op.conv2d)
```

Which means it will only do the casting when the specific architecture features are available (`has_asimd` will become `target.features.has_asimd` after https://github.com/apache/tvm/pull/12454 so is fine temporarily)

Reply:
That makes a lot of sense to me, and is definitely cleaner than special casing Cortex-M. To clarify though - are you suggesting removing the checks for `is_depthwise` and whether `attrs["data_layout"] == "NHWC"`?

Reply:
Good spot, I think we should keep the check as `(not is_depthwise) and has_asimd and attrs["data_layout"] == "NHWC"` for invoking the helper with the casting?

Reply:
Sounds good to me.

Reply:
I think you're right, my boolean logic was off methinks, as I remember it the logic should cast for ASIMD and opt out if there's another option which might work better:

```
    use_int8_on_arm = (not is_depthwise) and attrs["data_layout"] == "NHWC"
    has_dotprod = is_fast_int8_on_arm()
    other_options = use_int8_on_arm or has_dotprod
    if has_asimd() and not other_options:
        return helper_no_fast_int8_hw_legalization(attrs, inputs, types, relay.nn.conv2d)
    return helper_change_dtypes_to_be_same(attrs, inputs, types, relay.qnn.op.conv2d)
```

Does that sound right to you? 😸

Reply:
Yep, that sounds good, and should make us pass the existing unit tests. I think we should be good to merge once the CI is green.